man/man1/logtalk_allure_report.1
.TH logtalk_allure_report 1 "December 29, 2024" "Logtalk 3.87.0" "Logtalk Documentation"

.SH NAME
logtalk_allure_report \- Logtalk script for generating Allure reports

.SH SYNOPSIS
.B logtalk_allure_report -v
.br
.B logtalk_allure_report
[\fB-d \fItests\fR]
[\fB-i \fIresults\fR]
[\fB-o \fIreport\fR]
[\fB-t \fItitle\fR]
[\fB-p\fR]
[\fB-s\fR]
[\fB-- \fIenvironment pairs\fR]
.br
.B logtalk_allure_report -h

.SH DESCRIPTION
\fBlogtalk_allure_report\fR is a script for generating Allure reports (version 2.26.0 or later). Requires running first the \fIlogtalk_tester\fR script using the \fI-f xunit\fR (or \fI-f xunit_net_v2\fR) command-line option.

.SH OPTIONS
.TP
.BI \-v
Print script version.
.TP
.BI \-d
The directory containing the tests (i.e., the directory where the \fIlogtalk_tester\fR script was run; default is the current directory).
.TP
.BI \-i
Directory where to collect the test results (i.e., the xUnit reports) for generating the report (default is \fI./allure-results\fR).
.TP
.BI \-o
Directory where the report is to be saved (default is \fI./allure-report\fR). If the directory already exists and contains a previous report, its contents are deleted before generating the new report. If the directory exists, is not empty, and does not appear to contain a previous report, the script execution is terminated to prevent data loss.
.TP
.BI \-t
Report title (default is "Allure Report"; this option is ignored when using the \fI-p\fR option).
.TP
.BI \-p
Preprocess results but do not generate a report. Useful when the actual generation of the report is delegated to, e.g., a Docker service or a Jenkins plug-in.
.TP
.BI \-s
Generate a single file report.
.TP
.BI \--
Environment pairs (key1=value1 key2=value2 ...). Used to generate an \fIenvironment.properties\fR for adding additional information to the generated report.
.TP
.B \-h
Provide help on script usage.

.SH "EXIT STATUS"
.TP
.B 0
Allure report successfully created.
.TP
.B 1
Either the results directory does not exist or the report directory exists but it is not empty and does not contain a previous report.

.SH EXAMPLES
.TP
Run all tests in the current directory and sub-directories using the GNU Prolog backend (saving the test results in the required xUnit format), generate the Allure report, and open the report for browsing:
\fBlogtalk_tester -p gnu -f xunit\fR
.br
\fBlogtalk_allure_report\fR
.br
\fBallure open\fR
.PP
.TP
Set the title for the report and information on the backend used:
\fBlogtalk_allure_report -t "My application tests report" -- Backend='GNU Prolog' Commit=f7a106e\fR
.br
\fBallure open\fR
.PP
.TP
Save the processed test results and the report data to permanent directories so that the report shows tests run trends:
\fBlogtalk_allure_report -i "$HOME/my_project/allure-results" -o "$HOME/my_project/allure_report"\fR
.br
\fBallure open "$HOME/my_project/allure_report"\fR

.SH IMPLEMENTATION NOTES
Tested using Allure 2.30.0.

.SH "SEE ALSO"
The Allure website at https://docs.qameta.io/allure-report/
.PP
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2024 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2025 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.87.0 - January ??, 2025
=========================

Logtalk compiler and runtime
----------------------------

* ADDED: New `includes/2`, `object/3`, `protocol/3`, and `category/3`
properties to the `logtalk::loaded_file_property/2` predicate.

Documentation
-------------

* UPDATED: Notes on the Prolog integration scripts.

* FIXED: Grammar and spelling typos in scripts and tests documentation.

Tools
-----

* REMOVED: Deprecated `lgtdoc` tool `.js` scripts and `ie50.xsl` file.

* IMPROVED: Computation of the Halstead complexity metrics by considering
recursive predicate calls.

* FIXED: The `packs` tool `lgtenv.ps1` script when using the `-p` option.

Installers and installation scripts
-----------------------------------

* UPDATED: Synced the MacPorts portfile with the version used upstream.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* ADDED: VSCode support for finding dynamic predicate references in clauses
that assert or retract clauses for them.

* FIXED: VSCode support for finding predicate implementations when the
definition is found in an included file.

* FIXED: VSCode support for finding predicate references when the call is
found in an included file.

* FIXED: VSCode support for finding predicate declarations when the reference
is found in an included file.

* FIXED: VSCode support for navigating call hierarchies where an exception
would be thrown in some cases.

Installers and installation scripts
-----------------------------------

* REMOVED: Deprecated `logtalk_user_setup.js` script. 


3.86.0 - December 18, 2024
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Experimental `logtalk::message_prefix_file/6` hook predicate to
the message printing mechanism.

* FIXED: False left-recursion warning when compiling a grammar rule for
a multifile non-terminal that calls a local non-terminal with the same
indicator.

Prolog adapter and integration files
------------------------------------

* UPDATED: Ciao Prolog adapter file to pretend that modules are supported
to allow loading libraries and examples that require file `use_module/1`
directives.

Documentation
-------------

* ADDED: Notes on encrypting applications using XVM.

* FIXED: Multiple Handbook grammar and spelling typos.

* FIXED: Multiple man pages grammar and spelling typos.

Library
-------

* FIXED: The `os::sleep/1` predicate for XSB to also accept floats.

Tools
-----

* CHANGED: The `lgtunit` tool test compilation warning and error messages to
include the file context for better integration with other developer tools.

* CHANGED: The `diagrams` tool now uses a Unicode glyph for the magnifying
lens instead of a PNG image.

* CHANGED: The `diagrams` tool now uses a Graphviz `box` shape for Prolog
modules instead of a `tab`, which is now only used for directories and
libraries.

* REMOVED: Deprecated `lgt2svg.js` script from the `diagrams` tool.

* ADDED: Experimental `diagrams` tool support for the d2 graph language.

* UPDATED: The `diagrams` tool to support graph languages such as d2 where the
node identifiers are not global.

* UPDATED: The `diagrams` tool to also support the PearAI, Windsurf, PyCharm,
and IDEA text editors when generating diagrams with local file links.

* UPDATED: The `tutor` tool to also explain duplicated predicate/non-terminal
reference linter warnings.

* UPDATED: The `tutor` tool to call the new `logtalk::message_prefix_file/6`
hook predicate to support VSCode displaying warnings and errors augmented with
explanations.

* UPDATED: The `help` tool for the new `logtalk::message_prefix_file/6` hook
predicate.

* FIXED: The `diagrams` tool to also generate an `instance/class` caption for
metaclasses.

* FIXED: Bug in the `diagrams` tool where repeated nodes could be created for
external entities.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would miss
"includes" edges.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would show
duplicated "includes" edges.

* FIXED: The `diagrams` tool support for the graph languages to quote atoms
in predicate labels and predicate node contents when required.

* FIXED: The `diagrams` tool exclude options to allow generating external
nodes for the excluded directories, files, libraries, and entities.

* FIXED: Bug in the `lgtdoc` tool when printing a warning on missing entity
directive `info/1` key.

Examples
--------

* ADDED: New `permutations` example providing benchmarks based on list
permutations. Adapted from Prolog code written and contributed by Paul
Tarau.

* UPDATED: The `assumptions` example with another individual example.

Tests
-----

* ADDED: Additional tests for the left-recursion linter warning.

* ADDED: Additional tests for the Prolog standard `round/1` evaluable.

Installers and installation scripts
-----------------------------------

* IMPROVED: The manual installation script and MacPorts Portfile instructions
for defining the `LOGTALKHOME` and `LOGTALKUSER` environment variables.

IDEs, text editors, and syntax highlighters support

scripts/logtalk_allure_report.ps1
#############################################################################
## 
##   Allure report generator script
##   Last updated on August 2, 2024
## 
##   This file is part of Logtalk <https://logtalk.org/>  
##   SPDX-FileCopyrightText: 1998-2025 Paulo Moura <pmoura@logtalk.org>
##   SPDX-License-Identifier: Apache-2.0
##   
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##   
##       http://www.apache.org/licenses/LICENSE-2.0
##   
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
## 
#############################################################################


#Requires -Version 7.3

[CmdletBinding()]
param(
	[Parameter()]
	[String]$d = $pwd,
	[String]$i = "allure-results",
	[String]$o = "allure-report",
	[String]$t,
	[Switch]$p,
	[Switch]$s,
	[String]$e = "",
	[Switch]$v,
	[Switch]$h
)

Function Write-Script-Version {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve
	Write-Output ($myName + " 0.13")
}

Function Write-Usage-Help() {
	$myFullName = $MyInvocation.ScriptName
	$myName = Split-Path -Path $myFullName -leaf -Resolve 

	Write-Output ""
	Write-Output "This script generates Allure reports from test results."
	Write-Output ""
	Write-Output "Usage:"
	Write-Output ("  " + $myName + " [-d tests] [-i results] [-o report] [-t title] [-p] [-e pairs]")
	Write-Output ("  " + $myName + " -v")
	Write-Output ("  " + $myName + " -h")
	Write-Output ""
	Write-Output "Optional arguments:"
	Write-Output "  -d tests directory (default is the current directory)"
	Write-Output "  -i results directory (default is .\allure-results)"
	Write-Output "  -o report directory (default is .\allure-report)"
	Write-Output "  -t report title (default is `"Allure Report`")"
	Write-Output "  -p preprocess results but do not generate report"
	Write-Output "  -s single file report"
	Write-Output "  -e environment pairs as a string ('key1=value1,key2=value2,...')"
	Write-Output "  -v print version"
	Write-Output "  -h help"
	Write-Output ""
}

Function Check-Parameters() {
	if ($v -eq $true) {
		Write-Script-Version
		Exit
	}
	if ($h -eq $true) {
		Write-Usage-Help
		Exit
	}
}

###################### here it starts ############################ 

$minimal_version="2.26.0" 

if ((Get-Command "allure" -ErrorAction SilentlyContinue) -eq $null)  { 
	Write-Output "Error: allure is not installed!"
	Exit 1
} else {
	if ((allure --version) -lt $minimal_version) {
		Write-Output ("Warning: allure " + $minimal_version + " or later version is recommended!")
	}
}

Check-Parameters

if (Test-Path $o -PathType container) {
	if ((Test-Path $o\data) -and (Test-Path $o\export) -and (Test-Path $o\history) -and `
		(Test-Path $o\plugins) -and (Test-Path $o\widgets) -and (Test-Path $o\app.js) -and `
		(Test-Path $o\favicon.ico) -and (Test-Path $o\index.html) -and (Test-Path $o\styles.css)) {
		Write-Output "Warning: Overriding previous report..."
	} else {
		Write-Output "Error! Specified report directory is not empty and does not contain a previous"
		Write-Output "       report. Terminating the script execution to prevent any data loss."
		Exit
	}
}

# move all xunit_report.xml files found in the current directory
# and sub-directories to the $results directory; the files are
# renamed using an integer counter to avoid overwriting them

New-Item -Path $i -ItemType directory -Force > $null
try {
	Remove-Item -Path $i\xunit_report_*.xml -Recurse -Force
} catch {
	Write-Output "Error occurred at cleanup previous results"
}

$counter=0
Get-ChildItem -Path $d\xunit_report.xml -Recurse |
Foreach-Object {
	$counter++
	Move-Item -Path $_.FullName -Destination ($i + "\xunit_report_" + $counter + ".xml")
}

if ($e -ne "") {
	$e.Split(",") | Set-Content -Path $i/environment.properties
}

if ($p -eq $true) {
	Exit
}

# assume that the $i directory is kept between test
# runs and use a custom file to track the build order
if (Test-Path $i\history ) {
	$current_build = [int](Get-Content -Path $i\history\logtalk_build_number)
} else {
	$current_build = 1
}

# move the history from the previous report to the
# $results directory so that we can get trend graphs
if (Test-Path $o\history) {
	if (Test-Path $i\history) {
		try {
			Remove-Item -Path $i\history -Confirm -Recurse -Force
		} catch {
			Write-Output "Error occurred at cleanup results history"
		}
	}
	Move-Item -Path $o\history -Destination $i\history
	$next_build = $current_build + 1
	Set-Content -Path $i\history\logtalk_build_number -Value $next_build
} else {
	$next_build = $current_build
}

# add a minimal executor.json so that trend graphs show build labels
New-Item -Path $o -Name executor.json -ItemType "file" -Force > $null
Add-Content -Path $o\executor.json -Value ('"buildOrder": "' + $next_build + '"')
Add-Content -Path $o\executor.json -Value ('"buildName": "logtalk_allure_report#' + $next_build + '"')
Add-Content -Path $o\executor.json -Value '"name": "logtalk_tester"'
Add-Content -Path $o\executor.json -Value '"type": "logtalk_tester"'

# add minimal categories.json to classify failed tests
New-Item -Path $o -Name categories.json -ItemType "file" -Force > $null
Add-Content -Path $o\categories.json -Value '['
Add-Content -Path $o\categories.json -Value '	{'
Add-Content -Path $o\categories.json -Value '		"name": "Failed tests",'
Add-Content -Path $o\categories.json -Value '		"matchedStatuses": ["failed"]'
Add-Content -Path $o\categories.json -Value '	}'
Add-Content -Path $o\categories.json -Value ']'

Push-Location (Join-Path $i ..)

if ($s -eq $true) {
	if ($t -ne "") {
		allure generate --single-file --clean --name $t --report-dir $o $i
	} else {
		allure generate --single-file --clean --report-dir $o $i
	}
} else if ($t -ne "") {
	allure generate --clean --name $t --report-dir $o $i
} else {
	allure generate --clean --report-dir $o $i
}

Pop-Location
